src/gui/plugins/align_tool/AlignTool.cc

```diff
@@ -0,0 +1,647 @@
+  std::transform(newTarget.begin(), newTarget.end(),
+                 newTarget.begin(), ::tolower);
+
+  if (newTarget == "first")
+  {
+    this->dataPtr->first = true;
+  }
+  else if (newTarget == "last")
+  {
+    this->dataPtr->first = false;
+  }
+  else
+  {
+    ignwarn << "Invalid align target string: " << newTarget << "\n";
+    ignwarn << "The valid options are:\n";
+    ignwarn << " - first\n";
+    ignwarn << " - last\n";
+  }
+}
+
+/////////////////////////////////////////////////
+void AlignTool::OnHoveredEntered()
+{
+  std::lock_guard<std::mutex> lock(this->dataPtr->mutex);
+  this->AddState(AlignState::HOVER);
+}
+
+/////////////////////////////////////////////////
+void AlignTool::OnHoveredExited()
+{
+  std::lock_guard<std::mutex> lock(this->dataPtr->mutex);
+  // Always reset after an exit, if an align call is made, it
+  // will have updated the prev positions of the entities so a
+  // reset after align will result in no change
+  this->AddState(AlignState::RESET);
+}
+
+/////////////////////////////////////////////////
+void AlignTool::OnAlign()
+{
+  this->AddState(AlignState::ALIGN);
+}
+
+/////////////////////////////////////////////////
+void AlignTool::AddState(const AlignState &_state)
+{
+  this->dataPtr->states.push(_state);
+}
+
+////////////////////////////////////////////////
+void AlignTool::MakeTransparent(const rendering::NodePtr &_node)
+{
+  if (!_node)
+    return;
+
+  for (auto n = 0u; n < _node->ChildCount(); ++n)
+  {
+    auto child = _node->ChildByIndex(n);
+    this->MakeTransparent(child);
+  }
+
+  auto vis = std::dynamic_pointer_cast<rendering::Visual>(_node);
+  if (nullptr == vis)
+    return;
+
+  // Visual material
+  auto visMat = vis->Material();
+  if (nullptr != visMat)
+  {
+    // If the entity isn't already transparent, make it transparent
+    if (this->dataPtr->originalTransparency.find(vis->Name()) ==
+        this->dataPtr->originalTransparency.end())
+    {
+      this->dataPtr->originalTransparency[vis->Name()] = visMat->Transparency();
+      visMat->SetTransparency(visMat->Transparency() + 0.5);
+    }
+  }
+
+  for (auto g = 0u; g < vis->GeometryCount(); ++g)
+  {
+    auto geom = vis->GeometryByIndex(g);
+    auto wireBox = std::dynamic_pointer_cast<rendering::WireBox>(geom);
+
+    // Skip wirebox geometry
+    if (wireBox)
+      continue;
+
+    // Geometry material
+    auto geomMat = geom->Material();
+    if (nullptr == geomMat)
+      continue;
+
+    // If the entity isn't already transparent, make it transparent
+    if (this->dataPtr->originalTransparency.find(geom->Name()) ==
+        this->dataPtr->originalTransparency.end())
+    {
+      this->dataPtr->originalTransparency[geom->Name()] =
+          geomMat->Transparency();
+      geomMat->SetTransparency(geomMat->Transparency() + 0.5);
+    }
+  }
+}
+
+/////////////////////////////////////////////////
+void AlignTool::MakeSolid(const rendering::NodePtr &_node)
```

Comment:
`MakeOpaque`? 

The recursive code here is quite similar to `MakeTransparent` so you consider merging the two functions

Reply:
Resolved in [afe5d75b](https://github.com/ignitionrobotics/ign-gazebo/pull/59/commits/afe5d75b8f2ee1b8b38247903c54b5d596f02760)